### ISSUE TITLE
Increased RAM usage in v6.x provider series

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi-aws#3020]
### What happened?

Investigating a related performance regression I can confirm that v6 uses more RAM than v5. For example, provisioning 1000 KMS keys in the following program:

```python
import pulumi
import pulumi_aws as aws

for i in range(0, 1000):
    aws.kms.Key(f"key-{i}",
                deletion_window_in_days=7,
                description=f"KMS key {i}")
```

On v6.7.0 I am observing 300-400MB RAM peak aws-provider use.

On 5.42.0 the same program has 230MB RAM peak. 

The statefile sizes are about the same.

### Example

See above.

### Output of `pulumi about`

CLI          
Version      3.92.0
Go Version   go1.21.3
Go Compiler  gc

Plugins
NAME    VERSION
aws     5.42.0
python  unknown

Host     
OS       darwin
Version  14.0
Arch     x86_64

This project is written in python: executable='/Users/t0yv0/bin/python3' version='3.10.13'

Current Stack: t0yv0/repro-aws-lbriggs/dev

TYPE                  URN
pulumi:pulumi:Stack   urn:pulumi:dev::repro-aws-lbriggs::pulumi:pulumi:Stack::repro-aws-lbriggs-dev
pulumi:providers:aws  urn:pulumi:dev::repro-aws-lbriggs::pulumi:providers:aws::prov
pulumi:providers:aws  urn:pulumi:dev::repro-aws-lbriggs::pulumi:providers:aws::default_5_42_0
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-1
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-0
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-2
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-6
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-5
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-3
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-7
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-4
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-9
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-10
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-11
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-12
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-13
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-14
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-15
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-16
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-8
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-17
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-19
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-18
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-21
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-22
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-23
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-20
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-26
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-25
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-24
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-29

aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-958
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-959
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-960
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-961
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-962
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-963
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-964
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-965
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-966
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-967
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-968
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-969
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-970
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-971
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-972
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-973
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-974
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-975
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-976
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-977
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-978
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-979
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-980
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-981
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-982
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-983
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-984
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-985
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-986
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-987
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-988
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-989
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-990
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-991
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-992
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-993
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-994
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-995
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-996
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-997
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-998
aws:kms/key:Key       urn:pulumi:dev::repro-aws-lbriggs::aws:kms/key:Key::key-999

Found no pending operations associated with dev

Backend        
Name           pulumi.com
URL            https://app.pulumi.com/t0yv0
User           t0yv0
Organizations  t0yv0, pulumi
Token type     personal

Dependencies:
NAME        VERSION
pip         23.3.1
pulumi-aws  5.42.0
setuptools  68.2.2
wheel       0.41.3

Pulumi locates its logs in /var/folders/gk/cchgxh512m72f_dmkcc3d09h0000gp/T/ by default

### Additional context

Increased RAM usage can translate to slow operation in RAM-constrained environments such as deployment containers. The problem is exacerbated when multiple copies of the provider binary are in use. It is likely that Plugin Framework related metadata and initialization contributes to increased RAM usage but this needs confirmation.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by t0yv0** [comment]
Customer reporting additional impact, "the lack of logs and hanging seem to indicate an OOM".

I believe recent changes to the bridge and provider could help here a little bit, I'll give it another close look to measure.

https://github.com/pulumi/pulumi-terraform-bridge/issues/1454 should reduce the init pressure from processing automatic aliasing and renaming via JSON maps, this has not shipped yet.

https://github.com/pulumi/pulumi-aws/pull/3070 should help with startup allocations as well as wall-clock time performance (where it was primarily aimed at). I think 3070 has shipped already though (v6.15.0).

**Comment by t0yv0** [comment]
As an aside, setting `--parallel 20` to Pulumi greatly reduces the RAM consumption of Pulumi CLI (from 8GB to 50mb). But even with that I can still repro this issue on latest pulumi-aws. There seems to be a subtle memory leak where repeatedly provisioning resources creeps up the RAM usage. I will continue investigating.

**Comment by t0yv0** [comment]
Some progress, getting to previous levels of RAM usage with https://github.com/pulumi/pulumi-aws/pull/3172

**Comment by t0yv0** [comment]
Fixed in v6.17.0. Filed a proposal to set default parallelism in Pulumi CLI: https://github.com/pulumi/pulumi/issues/14989
